File: src/madmolecule.cpp
#include "madmolecule.hpp"

using namespace madness;

MadMolecule::MadMolecule() {
    mol = madness::Molecule();
}

void MadMolecule::add_atom(float pos_x, float pos_y, float pos_z, std::string symbol) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    pos_x *= scale;
    pos_y *= scale;
    pos_z *= scale;
    mol.add_atom(pos_x, pos_y, pos_z, (double)symbol_to_atomic_number(symbol), symbol_to_atomic_number(symbol));
}

std::string MadMolecule::to_json() {
    return mol.to_json().dump();
}


SavedFct MadMolecule::compute_nuclear_derivative(MadnessProcess& mp, const int atom, const int axis) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    madchem::MolecularDerivativeFunctor func(mol, atom, axis);
    real_function_3d op = real_factory_3d(*(mp.world)).functor(func).truncate_on_project().truncate_mode(0);
    op = scale * op; // scale the function to Angstrom
    return SavedFct(op);
}

SavedFct MadMolecule::compute_second_nuclear_derivative(MadnessProcess& mp, const int atom, const int axis1,
                                                  const int axis2) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;                                     
    madchem::MolecularSecondDerivativeFunctor func(mol, atom, axis1, axis2);
    real_function_3d op = real_factory_3d(*(mp.world)).functor(func).truncate_on_project().truncate_mode(0);
    op = scale * scale * op; // scale the function to Angstrom
    return SavedFct(op);
}

double MadMolecule::nuclear_repulsion_derivative(const int atom, const int axis) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    return scale * mol.nuclear_repulsion_derivative(atom, axis);
}

double MadMolecule::nuclear_repulsion_second_derivative(const int atom1, const int atom2, const int axis1, const int axis2) {
    double scale = 1e-10 / madness::constants::atomic_unit_of_length;
    return scale * scale * mol.nuclear_repulsion_second_derivative(atom1, atom2, axis1, axis2);
}


SavedFct MadMolecule::get_vnuc(MadnessProcess& mp) {
    PotentialManager pm = PotentialManager(mol, "");
    pm.make_nuclear_potential(*(mp.world));
    real_function_3d Vnuc = pm.vnuclear();
    return SavedFct(Vnuc);
}

File: src/madpy/madmolecule.py
from ._madpy_impl import MadMolecule as MadMoleculeImpl
import re

class MadMolecule:
    impl = None

    def __init__(self, geometry:str=None, *args, **kwargs):
        self.impl = MadMoleculeImpl(*args, **kwargs)
        if geometry is not None:
            geometry = geometry.lower()
            geometry = geometry.strip()
            # Replace tabs with spaces
            geometry = geometry.replace('\t', ' ')
            # Replace multiple whitespace characters with a single space
            geometry
            re.sub(r'\s+', ' ', geometry).strip()

            for line in geometry.split("\n"):
                data = line.split(" ")
                x = eval(data[1])
                y = eval(data[2])
                z = eval(data[3])
                s = data[0]
                self.add_atom(x, y, z, s)

    def add_atom(self, pos_x, pos_y, pos_z, symbol):
        self.impl.add_atom(pos_x, pos_y, pos_z, symbol)

    def to_json(self):
        return self.impl.to_json()

    def compute_nuclear_derivative(self, madworld, atom, axis):
        return self.impl.compute_nuclear_derivative(madworld._impl, atom, axis)
    
    def compute_second_nuclear_derivative(self,madworld, atom: int, axis1: int, axis2: int):
        return self.impl.compute_second_nuclear_derivative(madworld._impl, atom, axis1, axis2)

    def nuclear_repulsion_derivative(self, atom: int, axis: int):
        return self.impl.nuclear_repulsion_derivative(atom, axis)
        
    def nuclear_repulsion_second_derivative(self, atom1: int, atom2: int, axis1: int, axis2: int):
        return self.impl.nuclear_repulsion_second_derivative(atom1, atom2, axis1, axis2)


    def get_vnuc(self, madworld):
        return self.impl.get_vnuc(madworld._impl)

    def get_nuclear_charge(self):
        return self.impl.get_nuclear_charge()

    def get_nuclear_repulsion(self):
        return self.impl.get_nuclear_repulsion()

    @property
    def n_electrons(self):
        return int(self.get_nuclear_charge())

    @property
    def n_core_electrons(self):
        return self.impl.get_core_n_electrons()
